### DISCUSSION TITLE
load sprite json with custom protocol does not JSON parse ArrayData

### DISCUSSION THREAD

**Discussion by halset** [maplibre/maplibre-gl-js#5003] [Q&A]
**maplibre-gl-js version**: main branch

**browser**: Safari

### Steps to Trigger Behavior

 1. Add custom protocol using `maplibregl.addProtocol('internal'...`
 2. Have a style with a sprite using that protocol.
 3. Observe that sprite is fetched, but that none of the icons can be used in the map.

### Link to Demonstration

https://jsbin.com/jusasoy/5/edit?html,output

### Expected Behavior

See the map with 3 red circles that each have an airplane icon in them.
<img width="226" alt="Skjermbilde 2024-11-05 kl  12 29 42" src="https://github.com/user-attachments/assets/c5a9026a-d598-4409-9cad-139421b91936">

### Actual Behavior

Can see map with 3 red circles, but without airplane icons.
<img width="258" alt="Skjermbilde 2024-11-05 kl  12 29 15" src="https://github.com/user-attachments/assets/e0e4148c-4870-4dec-ac17-a66b5f56dac8">

**Comment by halset** [comment]
Here is a patch that works, but I do not know if this is the place to solve this.

```patch
diff --git a/src/style/load_sprite.ts b/src/style/load_sprite.ts
index 391317442..6252c4c9a 100644
--- a/src/style/load_sprite.ts
+++ b/src/style/load_sprite.ts
@@ -64,7 +64,14 @@ async function doOnceCompleted(
         result[spriteName] = {};
 
         const context = browser.getImageCanvasContext((await imagesMap[spriteName]).data);
-        const json = (await jsonsMap[spriteName]).data;
+        const jsonResponseData = (await jsonsMap[spriteName]).data;
+
+        var json
+        if (jsonResponseData instanceof ArrayBuffer) {
+            json = JSON.parse(new TextDecoder().decode(jsonResponseData as ArrayBuffer));
+        } else {
+            json = jsonResponseData;
+        }
 
         for (const id in json) {
             const {width, height, x, y, sdf, pixelRatio, stretchX, stretchY, content, textFitWidth, textFitHeight} = json[id];

```

**Comment by halset** [comment]
It is also possible to work around this problem by parsing the JSON in the `addProtocol` callback. 

```js
    // internal:// links handled generally, but not by maplibre. this is an attempt to work around that problem
    maplibregl.addProtocol('internal', async (params, abortController) => {
        const t = ...;
        if (t.status == 200 || t.status == 0) {
            const buffer = await t.arrayBuffer();
            
            // work around https://github.com/maplibre/maplibre-gl-js/issues/5002
            if (params.url.includes("sprite") && params.url.endsWith(".json")) {
                const parsedJSON = JSON.parse(new TextDecoder().decode(buffer));
                return {'data': parsedJSON}
            }
            
            return {'data': buffer}
        } else {
            throw new Error('got status: ' + t.status + ' ' + t.statusText + ' for: ' + params.url);
        }
    });

```

**Comment by HarelM** [comment]
You need the `addProtocol` to act differently based on the request type, hare's a hack I did just now:
```ts
maplibregl.addProtocol('internal', async (params, abortController) => {
        const fixedUrl = params.url.replace('internal://', 'https://')
        const t = await fetch(fixedUrl);
        if (t.status == 200 || t.status == 0) {
          if (fixedUrl.includes(".json")) {
            const text = await t.text();
            return {
                'data': JSON.parse(text)
            }
          } else {
            const buffer = await t.arrayBuffer();
            return {
                'data': buffer
            }
          }
        } else {
            throw new Error('got status: ' + t.status + ' ' + t.statusText + ' for: ' + params.url);
        }
    });
```

**Reply by halset** [reply]
Yes, I am using that workaround now as seen in https://github.com/maplibre/maplibre-gl-js/discussions/5003#discussioncomment-11154019. 

This was very confusing for me. Seems like raster tiles, vector tiles and sprite images can be returned as raw data, but json has to be parsed. 

It would make more sense to me if json also could be returned as raw data. Perhaps it would make more sense to other new users to custom protocol handling as well. 

However, with the workaround, I have a working solution. So it is not that important to fix for me.
